File: 백준/Silver/7562. 나이트의 이동/나이트의 이동.py
# 아이디어
# bfs거리 계산
# 나이트로 시작점에서 목표점까지 최단 거리 구하기

from collections import deque

dr = [-2, -1, 1, 2, 2, 1, -1, -2]
dc = [1, 2, 2, 1, -1, -2, -2, -1]

T = int(input())
for t in range(1, T + 1):
    N = int(input())
    sr, sc = map(int, input().split())  # 시작점
    er, ec = map(int, input().split())  # 목표점

    visited = [[0] * N for _ in range(N)]  # 방문 처리 및 이동 횟수 저장

    queue = deque([[sr, sc]])
    visited[sr][sc] = 1

    while queue:
        q = queue.pop()
        cr, cc = q[0], q[1]
        if cr == er and cc == ec:  # 도착했으면
            break  # while탈출!
        for d in range(8):
            nr, nc = cr + dr[d], cc + dc[d]
            if 0 <= nr < N and 0 <= nc < N:
                if visited[nr][nc]:  # 이미 방문했으면
                    continue  # skip
                queue.appendleft([nr, nc])
                visited[nr][nc] = visited[cr][cc] + 1

    print(visited[er][ec] - 1 )

File: 백준/Silver/7562. 나이트의 이동/나이트의 이동.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.ArrayDeque;
import java.util.Arrays;
import java.util.Deque;
import java.util.Iterator;
import java.util.StringTokenizer;

public class Main {

	static int N, sr, sc, er, ec;
	static Deque<int[]> queue;
	static boolean[][] visited;
	static int[] dr = { -2, -1, 1, 2, 2, 1, -1, -2 };
	static int[] dc = { 1, 2, 2, 1, -1, -2, -2, -1 };

	public static void main(String[] args) throws Exception {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		int TC = Integer.parseInt(br.readLine());

		for (int i = 0; i < TC; i++) {
			N = Integer.parseInt(br.readLine());

			StringTokenizer st = new StringTokenizer(br.readLine());
			sr = Integer.parseInt(st.nextToken());
			sc = Integer.parseInt(st.nextToken());

			st = new StringTokenizer(br.readLine());
			er = Integer.parseInt(st.nextToken());
			ec = Integer.parseInt(st.nextToken());

			queue = new ArrayDeque<int[]>();
			visited = new boolean[N][N];

			// 입력 완료 로직 시작!!
			System.out.println(bfs());

		}

	}

	private static int bfs() {
		queue.offer(new int[] { sr, sc, 0 });
		visited[sr][sc] = true;

		while (!queue.isEmpty()) {
			int[] q = queue.poll();
			int cr = q[0];
			int cc = q[1];
			int ct = q[2];
			
			if(cr == er && cc == ec) {
				return ct;
			}

			for (int d = 0; d < 8; d++) {
				int nr = cr + dr[d];
				int nc = cc + dc[d];
				int nt = q[2] + 1;

				if (!check(nr, nc))
					continue;
				if (visited[nr][nc])
					continue;
				
				queue.offer(new int[] {nr, nc, nt});
				visited[nr][nc] = true;
			}
		}
		return -1;

	}

	private static boolean check(int r, int c) {
		return 0 <= r && r < N && 0 <= c && c < N;
	}

}
